File: rgn_concurrency/rgn_concurrency_akka_essentials/src/main/java/org/rgn/first/mapreduce/actors/ReduceActor.java
package org.rgn.first.mapreduce.actors;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.rgn.first.mapreduce.messages.MapData;
import org.rgn.first.mapreduce.messages.ReduceData;
import org.rgn.first.mapreduce.messages.WordCount;

import akka.actor.UntypedActor;

public class ReduceActor extends UntypedActor {

	@Override
	public void onReceive(Object message) throws Exception {
		if (message instanceof MapData) {
			MapData mapData = (MapData) message;
			getSender().tell(reduce(mapData.getDataList()), getSelf());

		} else {
			unhandled(message);
		}
	}

	private ReduceData reduce(List<WordCount> dataList) {
		Map<String, Integer> reduceMap = new HashMap<>();

		for (WordCount wordCount : dataList) {
			if (reduceMap.containsKey(wordCount.getWord())) {
				Integer value = reduceMap.get(wordCount.getWord());
				value++;
				reduceMap.put(wordCount.getWord(), value);

			} else {
				reduceMap.put(wordCount.getWord(), 1);
			}
		}

		return new ReduceData(reduceMap);
	}

}

File: rgn_concurrency/rgn_concurrency_akka_essentials/src/main/scala/org/rgn/scala/first/mapreduce/actors/ReduceActor.scala
package org.rgn.scala.first.mapreduce.actors

import akka.actor.Actor
import org.rgn.scala.first.mapreduce.MapData
import org.rgn.scala.first.mapreduce.WordCount
import org.rgn.scala.first.mapreduce.ReduceData

class ReduceActor extends Actor {

  def receive: Receive = {
    case MapData(dataList) =>
      sender ! reduce(dataList)
  }
  def reduce(words: IndexedSeq[WordCount]): ReduceData = ReduceData {
    words.foldLeft(Map.empty[String, Int]) {
      (index, words) =>
        if (index contains words.word)
          index + (words.word -> (index.get(words.word).get + 1))
        else
          index + (words.word -> 1)
    }
  }

}